### PR TITLE
ceph: add external support for object store user

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

Now, the object store user is capable of creating s3 users on an
external Ceph cluster.

Signed-off-by: Sébastien Han <seb@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/cluster/mon/config.go [hunk 1]
```diff
@@ -319,3 +322,42 @@ func ExtractKey(contents string) (string, error) {
...
+
+// PopulateExternalClusterInfo Add validation in the code to fail if the external cluster has no OSDs keep waiting
+func PopulateExternalClusterInfo(context *clusterd.Context, namespace string) *cephconfig.ClusterInfo {
+	var clusterInfo *cephconfig.ClusterInfo
+	for {
+		var err error
+		clusterInfo, _, _, err = LoadClusterInfo(context, namespace)
+		if err != nil {
+			logger.Warningf("waiting for the connection info of the external cluster. retrying in %s.", externalConnectionRetry.String())
+			time.Sleep(externalConnectionRetry)
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mon/config.go:334]: Instead of the `for` loop and sleeping forever, shouldn't we be re-queueing the event?

File: pkg/operator/ceph/object/user/controller.go [hunk 2]
```diff
@@ -168,13 +172,27 @@ func (r *ReconcileObjectStoreUser) reconcile(request reconcile.Request) (reconci
...
+	r.cephClusterSpec = &cephCluster.Spec
 
 	// Set a finalizer so we can do cleanup before the object goes away
 	err = opcontroller.AddFinalizerIfNotPresent(r.client, cephObjectStoreUser)
 	if err != nil {
 		return reconcile.Result{}, errors.Wrap(err, "failed to add finalizer")
 	}
 
+	// Populate clusterInfo
+	// Always populate it during each reconcile
```
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go:184]: We have to reload it every time in case it changes?

File: pkg/operator/ceph/controller/controller_utils.go [hunk 3]
```diff
@@ -67,27 +65,15 @@ func IsReadyToReconcile(c client.Client, clustercontext *clusterd.Context, names
...
-	// If the cluster is healthy
-	// Test a Ceph command to verify the Operator is ready
-	// This is done to silence errors when the operator just started and cannot reconcile yet
-	status, err := cephclient.Status(clustercontext, namespacedName.Namespace)
-	if err != nil {
-		if strings.Contains(err.Error(), "error calling conf_read_file") {
-			logger.Infof("%q: operator is not ready to run ceph command, cannot reconcile yet.", controllerName)
-			return cephCluster, false, cephClusterExists, WaitForRequeueIfCephClusterNotReady
+	// read the CR status of the cluster
+	if cephCluster.Status.CephStatus != nil {
```
Comment (travisn) [inline pkg/operator/ceph/controller/controller_utils.go:69]: Is this change needed for the object store user? Or it's just removing the unnecessary call to retrieve the status since it's already available?

File: pkg/operator/ceph/controller/controller_utils.go [hunk 4]
```diff
@@ -67,27 +65,15 @@ func IsReadyToReconcile(c client.Client, clustercontext *clusterd.Context, names
...
-		// We should not arrive there
-		logger.Errorf("%q: ceph command error %v", controllerName, err)
-		return cephCluster, false, cephClusterExists, ImmediateRetryResult
-	}
-
-	// If Ceph status is ok we can reconcile
-	if status.Health.Status == "HEALTH_OK" || status.Health.Status == "HEALTH_WARN" {
-		logger.Debugf("%q: ceph status is %q, operator is ready to run ceph command, reconciling", controllerName, status.Health.Status)
-		return cephCluster, true, cephClusterExists, reconcile.Result{}
+		logger.Infof("%s: CephCluster %q found but skipping reconcile since ceph health is %q", controllerName, cephCluster.Name, cephCluster.Status.CephStatus)
```
Comment (travisn) [inline pkg/operator/ceph/controller/controller_utils.go:74]: This is for a separate discussion, but I still don't see why we block reconciling even if the ceph health has an error. For example, if mds raises a health error, it shouldn't block the reconcile of object stores or object users.

File: pkg/operator/ceph/object/user/controller.go (hunk 2 above)
Comment (leseb) [reply pkg/operator/ceph/object/user/controller.go:184]: Yes, I'd say better safe than sorry, e,g, if the mon changes.

File: pkg/operator/ceph/object/user/controller.go [hunk 5]
```diff
@@ -356,38 +386,50 @@ func (r *ReconcileObjectStoreUser) reconcileCephUserSecret(cephObjectStoreUser *
...
 	}
 	logger.Debug("CephObjectStore exists")
 
-	pods, err := r.getRgwPodList(cephObjectStoreUser)
-	if err != nil {
-		return err
-	}
+	// There are no pods running when the cluster is external
+	// Unless you pass the admin key...
+	if !r.cephClusterSpec.External.Enable {
```
[the remark this answers is not in the document]
Comment (leseb) [reply pkg/operator/ceph/object/user/controller.go:397]: Sorry, I mixed this with some code for 1.4, since in 1.4 we would actually get the CephObjectStore. For now, let's just return at L314, much simpler.

File: pkg/operator/ceph/cluster/mon/config.go (hunk 1 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/mon/config.go:334]: I'm just re-using the old code, I cannot quite do this change since it will affect the CephCluster CR too, and in 1.3 it's not controller runtime yet.

File: pkg/operator/ceph/controller/controller_utils.go (hunk 4 above)
Comment (leseb) [reply pkg/operator/ceph/controller/controller_utils.go:74]: That's a good point actually, I'm fine letting the controller reconcile though. It's probably what we should do anyway. Follow-up PR in master?

File: pkg/operator/ceph/controller/controller_utils.go (hunk 3 above)
Comment (leseb) [reply pkg/operator/ceph/controller/controller_utils.go:69]: It's needed since when we do an external cluster we cannot quite call the `Status` function. If we want to do this, we would need to instruct this function which user to use so that we can properly connect to the cluster. See we have `Status` and `StatusWithUser()`

I thought that relying on the CR status was good enough.

**Comment by leseb** [comment]
> @leseb : Just out of curiosity, do you think a similar change is required for obc as well?

Nop, OBC has that support already.

File: pkg/operator/ceph/object/user/controller.go [hunk 6]
```diff
@@ -452,13 +452,9 @@ func (r *ReconcileObjectStoreUser) getRgwPodList(cephObjectStoreUser *cephv1.Cep
 
 // Delete the user
 func (r *ReconcileObjectStoreUser) deleteUser(u *cephv1.CephObjectStoreUser) error {
-	_, rgwerr, err := object.DeleteUser(r.objContext, u.Name)
+	output, err := object.DeleteUser(r.objContext, u.Name)
 	if err != nil {
-		if rgwerr == 2 {
-			logger.Infof("ceph object user %q does not exist in store %q", u.Name, u.Spec.Store)
```
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go:420]: If the user doesn't exist, don't we still need to ignore this error?

File: pkg/util/exec/exec.go [hunk 7]
```diff
@@ -269,7 +268,7 @@ func runCommandWithOutput(cmd *exec.Cmd, combinedOutput bool) (string, error) {
 	out := strings.TrimSpace(string(output))
 
 	if err != nil {
-		return out, errors.Wrapf(err, "%s", string(err.(*exec.ExitError).Stderr))
+		return fmt.Sprintf("%s. %s", out, string(err.(*exec.ExitError).Stderr)), err
```
Comment (travisn) [inline pkg/util/exec/exec.go:271]: if combinedOutput is true, the stderr will already be included in the `out` var. The `cmd.CombinedOutput()` call on line 263 will return it.

Comment (leseb) [reply pkg/util/exec/exec.go:271]: Ok so I should do this only in L265 right?

File: pkg/operator/ceph/object/user/controller.go (hunk 6 above)
Comment (leseb) [reply pkg/operator/ceph/object/user/controller.go:420]: If the user does not exist `DeleteUser` returns nil now.

File: pkg/util/exec/exec.go [hunk 8]
```diff
@@ -259,17 +258,19 @@ func logOutput(stdout, stderr io.ReadCloser) {
 func runCommandWithOutput(cmd *exec.Cmd, combinedOutput bool) (string, error) {
 	var output []byte
 	var err error
+	var out string
 
 	if combinedOutput {
 		output, err = cmd.CombinedOutput()
 	} else {
 		output, err = cmd.Output()
+		out = fmt.Sprintf("%s. %s", string(output), string(err.(*exec.ExitError).Stderr))
```
Comment (travisn) [inline pkg/util/exec/exec.go:267]: the stderr can only be appended if err != nil

File: pkg/operator/ceph/object/user/controller.go (hunk 6 above)
Comment (travisn) [reply pkg/operator/ceph/object/user/controller.go:420]: good, nice that the third return code is removed as well
